### PR TITLE
Allow opening Realms with asymmetric tables

### LINKED ISSUE #1591: Can't open a Realm that has asymmetric tables
**Describe the bug**
Studio errors out when opening a Realm with asymmetric tables.

**To Reproduce**
1. Create a FLX Realm using any SDK.
2. Add an asymmetric model.
3. Try to open the Realm file using Studio.

**Expected behavior**
Studio opens and works fine.

**Screenshots**

Screenshot 2023-09-07 at 13 12 18

**Versions:**
 - OS: Windows
 - Studio Version: 14.0.3

**Additional context**
This is the stacktrace I get:

```
Error: You cannot query an asymmetric class.
    at getSchemaLength (file:///C:/Program%20Files/Realm%20Studio/resources/app.asar/build/realm-browser.renderer.bundle.js:1:512748)
    at file:///C:/Program%20Files/Realm%20Studio/resources/app.asar/build/realm-browser.renderer.bundle.js:1:500716
    at Array.map (<anonymous>)
    at ht (file:///C:/Program%20Files/Realm%20Studio/resources/app.asar/build/realm-browser.renderer.bundle.js:1:500168)
    at renderWithHooks (C:\Program Files\Realm Studio\resources\app.asar\node_modules\react-dom\cjs\react-dom.development.js:14803:18)
    at updateFunctionComponent (C:\Program Files\Realm Studio\resources\app.asar\node_modules\react-dom\cjs\react-dom.development.js:17034:20)
    at beginWork (C:\Program Files\Realm Studio\resources\app.asar\node_modules\react-dom\cjs\react-dom.development.js:18610:16)
    at HTMLUnknownElement.callCallback (C:\Program Files\Realm Studio\resources\app.asar\node_modules\react-dom\cjs\react-dom.development.js:188:14)
    at HTMLUnknownElement.sentryWrapped (C:\Program Files\Realm Studio\resources\app.asar\node_modules\@sentry\browser\cjs\helpers.js:92:17)
    at Object.invokeGuardedCallbackDev (C:\Program Files\Realm Studio\resources\app.asar\node_modules\react-dom\cjs\react-dom.development.js:237:16)
    at invokeGuardedCallback (C:\Program Files\Realm Studio\resources\app.asar\node_modules\react-dom\cjs\react-dom.development.js:292:31)
    at beginWork$1 (C:\Program Files\Realm Studio\resources\app.asar\node_modules\react-dom\cjs\react-dom.development.js:23203:7)
    at performUnitOfWork (C:\Program Files\Realm Studio\resources\app.asar\node_modules\react-dom\cjs\react-dom.development.js:22157:12)
    at workLoopSync (C:\Program Files\Realm Studio\resources\app.asar\node_modules\react-dom\cjs\react-dom.development.js:22130:22)
    at performSyncWorkOnRoot (C:\Program Files\Realm Studio\resources\app.asar\node_modules\react-dom\cjs\react-dom.development.js:21756:9)
    at C:\Program Files\Realm Studio\resources\app.asar\node_modules\react-dom\cjs\react-dom.development.js:11089:24
    at unstable_runWithPriority (C:\Program Files\Realm Studio\resources\app.asar\node_modules\scheduler\cjs\scheduler.development.js:653:12)
    at runWithPriority$1 (C:\Program Files\Realm Studio\resources\app.asar\node_modules\react-dom\cjs\react-dom.development.js:11039:10)
    at flushSyncCallbackQueueImpl (C:\Program Files\Realm Studio\resources\app.asar\node_modules\react-dom\cjs\react-dom.development.js:11084:7)
    at flushSyncCallbackQueue (C:\Program Files\Realm Studio\resources\app.asar\node_modules\react-dom\cjs\react-dom.development.js:11072:3)
    at scheduleUpdateOnFiber (C:\Program Files\Realm Studio\resources\app.asar\node_modules\react-dom\cjs\react-dom.development.js:21199:9)
    at Object.enqueueSetState (C:\Program Files\Realm Studio\resources\app.asar\node_modules\react-dom\cjs\react-dom.development.js:12639:5)
    at Component.setState (C:\Program Files\Realm Studio\resources\app.asar\node_modules\react\cjs\react.development.js:471:16)
    at zt.onRealmLoaded (file:///C:/Program%20Files/Realm%20Studio/resources/app.asar/build/realm-browser.renderer.bundle.js:1:509038)
    at zt.loadRealm (file:///C:/Program%20Files/Realm%20Studio/resources/app.asar/build/754.renderer.bundle.js:1:166602)
    in ht (created by wt)
    in wt (created by Bt)
    in div (created by Bt)
    in Bt (created by zt)
    in zt (created by G)
    in G
    in q
```

### PR DESCRIPTION
**Pull request by nirinchev**
Fixes https://github.com/realm/realm-studio/issues/1591

### CODE REVIEW

File: RELEASENOTES.md [hunk 1]
```diff
@@ -6,7 +6,7 @@
 
 ### Fixed
 
-- None
+- Fixed a bug where opening a Realm that contains asymmetric tables will crash the process with an error saying `You cannot query an asymmetric class`. ([#1591](https://github.com/realm/realm-studio/issues/1591))
```
Comment (elle-j) [inline RELEASENOTES.md:9]: I believe this should be moved to CHANGELOG instead?

File: src/ui/RealmBrowser/Content/AsymmetricObjectCollection.ts [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+  public description(): string {
+    throw new Error('Not implemented');
+  }
+
+  public isValid(): boolean {
+    return true;
+  }
+
+  public isEmpty(): boolean {
+    return this.length === 0;
```
Comment (elle-j) [inline src/ui/RealmBrowser/Content/AsymmetricObjectCollection.ts]: We could also return a hardcoded `true` here (and/or perhaps assert that the length is 0 if need be).

File: src/ui/RealmBrowser/Content/AsymmetricObjectCollection.ts [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+  public addListener() {
+    throw new Error('Not implemented');
+  }
+
+  public removeListener() {
+    throw new Error('Not implemented');
+  }
+
+  public removeAllListeners() {
```
Comment (elle-j) [inline src/ui/RealmBrowser/Content/AsymmetricObjectCollection.ts]: Saw that `SingleObjectCollection` does not have an explicit return types here, but good to add it 👍 
```suggestion
  public addListener(): void {
    throw new Error('Not implemented');
  }

  public removeListener(): void {
    throw new Error('Not implemented');
  }

  public removeAllListeners(): void {
```

File: src/ui/RealmBrowser/focus.ts [hunk 4]
```diff
@@ -43,7 +44,7 @@ interface IFocus {
 
 export interface IClassFocus extends IFocus {
   kind: 'class';
-  results: Realm.Results<any>;
+  results: Realm.Results<any> | AsymmetricObjectCollection<any>;
```
Comment (elle-j) [inline src/ui/RealmBrowser/focus.ts]: Don't think the union should be needed as it implements `Realm.Results`.
```suggestion
  results: Realm.Results<any>;
```

File: src/ui/RealmBrowser/index.tsx [hunk 5]
```diff
@@ -700,7 +705,8 @@ class RealmBrowserContainer
 
   private getSchemaLength = (name: string) => {
     if (this.realm) {
-      return this.realm.objects(name).length;
+      const schema = this.realm.schema.find(s => s.name == name);
```
Comment (elle-j) [inline src/ui/RealmBrowser/index.tsx]:
```suggestion
      const schema = this.realm.schema.find(s => s.name === name);
```

**Review by elle-j — APPROVED** [review]
Thanks for the fix Nikola!

File: src/ui/RealmBrowser/Content/AsymmetricObjectCollection.ts [hunk 6]
```diff
@@ -0,0 +1,94 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2020 Realm Inc.
```
Comment (elle-j) [inline src/ui/RealmBrowser/Content/AsymmetricObjectCollection.ts]:
```suggestion
// Copyright 2023 Realm Inc.
```

**Review by elle-j — APPROVED** [review]

**Review by kneth — APPROVED** [review]

**Pull request merged**